Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPILocalization.mm

```diff
@@ -0,0 +1,268 @@
+        @"browser.test.assertRejects(() => browser.i18n.getMessage(), /'message' argument is invalid because a string is expected./)",
+        @"browser.test.assertRejects(() => browser.i18n.getMessage(123), /'message' argument is invalid because a string is expected./)",
+        @"browser.test.assertRejects(() => browser.i18n.getMessage(null), /'message' argument is invalid because a string is expected./)",
+        @"browser.test.assertRejects(() => browser.i18n.getMessage(undefined), /'message' argument is invalid because a string is expected./)",
+
+        // Finish
+        @"browser.test.notifyPass()",
+    ]);
+
+    auto extension = adoptNS([[_WKWebExtension alloc] _initWithManifestDictionary:manifest resources:@{ @"background.js": backgroundScript, @"_locales/en/messages.json": messages }]);
+    auto manager = adoptNS([[TestWebExtensionManager alloc] initForExtension:extension.get()]);
+
+    // Set a base URL so it is a known value and not the default random one.
+    manager.get().context.baseURL = [NSURL URLWithString:baseURLString];
+
+    [manager loadAndRun];
+}
+
+TEST(WKWebExtensionAPILocalization, i18n)
+{
+    NSArray<NSString *> *preferredLocaleIdentifiers = NSLocale.preferredLanguages;
+    NSMutableOrderedSet<NSString *> *acceptedLanguages = [NSMutableOrderedSet orderedSetWithArray:preferredLocaleIdentifiers];
+    for (NSString *localeIdentifier in preferredLocaleIdentifiers)
+        [acceptedLanguages addObject:[NSLocale localeWithLocaleIdentifier:localeIdentifier].languageCode];
+
+    auto *acceptedLanguagesString = Util::constructJSArrayOfStrings(acceptedLanguages.array);
+
+    auto *backgroundScript = Util::constructScript(@[
+        // Variable setup
+        [NSString stringWithFormat:@"const acceptedLanguages = %@", acceptedLanguagesString],
+        [NSString stringWithFormat:@"const currentUILanguage = '%@'", currentLocaleString],
+        [NSString stringWithFormat:@"const textDirection = '%@'", textDirection],
+        [NSString stringWithFormat:@"const reversedTextDirection = '%@'", reversedTextDirection],
+        [NSString stringWithFormat:@"const startEdge = '%@'", startEdge],
+        [NSString stringWithFormat:@"const endEdge = '%@'", endEdge],
+
+        // Test predefined localization
+        @"browser.test.assertEq(await browser.i18n.getMessage('extension_name'), 'Web Extension Test Extension')",
+        @"browser.test.assertEq(await browser.i18n.getMessage('@@ui_locale'), 'en')",
+        @"browser.test.assertEq(await browser.i18n.getMessage('@@bidi_dir'), textDirection)",
+        @"browser.test.assertEq(await browser.i18n.getMessage('@@bidi_reversed_dir'), reversedTextDirection)",
+        @"browser.test.assertEq(await browser.i18n.getMessage('@@bidi_start_edge'), startEdge)",
+        @"browser.test.assertEq(await browser.i18n.getMessage('@@bidi_end_edge'), endEdge)",
+        @"browser.test.assertEq(await browser.i18n.getMessage('unknown_message'), '')",
```

Comment:
Shouldn't need `await` for `getMessage()` since it is a sync function that returns a result.